modal/_utils/pattern_matcher.py

```diff
@@ -0,0 +1,274 @@
+            pattern = os.path.normpath(pattern)
+            new_pattern = Pattern()
+            if pattern[0] == "!":
+                if len(pattern) == 1:
+                    raise ValueError('Illegal exclusion pattern: "!"')
+                new_pattern.exclusion = True
+                pattern = pattern[1:]
+                self.exclusions = True
+            # In Python, we can proceed without explicit syntax checking
+            new_pattern.cleaned_pattern = pattern
+            new_pattern.dirs = pattern.split(os.path.sep)
+            self.patterns.append(new_pattern)
+
+    def matches(self, file_path: str) -> bool:
+        """Check if the file path or any of its parent directories match the patterns.
+
+        This is equivalent to `MatchesOrParentMatches()` in the original Go
+        library. The reason is that `Matches()` in the original library is
+        deprecated due to buggy behavior.
+        """
+        matched = False
+        file_path = os.path.normpath(file_path)
+        if file_path == ".":
+            # Don't let them exclude everything; kind of silly.
+            return False
+        parent_path = os.path.dirname(file_path)
+        if parent_path == "":
+            parent_path = "."
+        parent_path_dirs = parent_path.split(os.path.sep)
+
+        for pattern in self.patterns:
+            # Skip evaluation based on current match status and pattern exclusion
+            if pattern.exclusion != matched:
+                continue
+
+            match = pattern.match(file_path)
+
+            if not match and parent_path != ".":
+                # Check if the pattern matches any of the parent directories
+                for i in range(len(parent_path_dirs)):
+                    dir_path = os.path.sep.join(parent_path_dirs[: i + 1])
+                    if pattern.match(dir_path):
+                        match = True
+                        break
+
+            if match:
+                matched = not pattern.exclusion
+
+        return matched
+
+
+def read_ignorefile(reader: TextIO) -> List[str]:
+    """Read an ignore file from a reader and return the list of file patterns to
+    ignore, applying the following rules:
+
+    - An UTF8 BOM header (if present) is stripped. (Python does this already)
+    - Lines starting with "#" are considered comments and are skipped.
+
+    For remaining lines:
+
+    - Leading and trailing whitespace is removed from each ignore pattern.
+    - It uses `os.path.normpath` to get the shortest/cleanest path for ignore
+      patterns.
+    - Leading forward-slashes ("/") are removed from ignore patterns, so
+      "/some/path" and "some/path" are considered equivalent.
+
+    Args:
+        reader (file-like object): The input stream to read from.
+
+    Returns:
+        list: A list of patterns to ignore.
+    """
+    if reader is None:
+        return []
+
+    excludes: List[str] = []
+
+    for line in reader:
+        pattern = line.rstrip("\n\r")
```

Comment:
Do we need this if we're going to `.strip()` anyway?

Reply:
Good point, it's not strictly necessary. But I think it aids in clarity since Python automatically includes the `\r\n` or `\n` at the end of the line when iterating through a file like this.